File: lib/cfr/builtins/libs/dart/ui/textBox.dart
import 'dart:core';
import 'dart:ui';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedTextBox extends VMManagedBox<TextBox> {
  VMManagedTextBox(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['left'] = vmObject.left;
    table['top'] = vmObject.top;
    table['right'] = vmObject.right;
    table['bottom'] = vmObject.bottom;
    table['direction'] = TextDirection.values.indexWhere((x) {
      return x == vmObject.direction;
    });
    table['toRect'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Rect>(
            object: vmObject.toRect(),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getStart'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.start,
      ];
    });
    table['getEnd'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.end,
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final TextBox vmObject;
}

void loadTextBox({required HydroState hydroState, required HydroTable table}) {
  table['textBoxFromLTRBD'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<TextBox>(
          object: TextBox.fromLTRBD(
              luaCallerArguments[1]?.toDouble(),
              luaCallerArguments[2]?.toDouble(),
              luaCallerArguments[3]?.toDouble(),
              luaCallerArguments[4]?.toDouble(),
              maybeUnBoxEnum(
                  values: TextDirection.values,
                  boxedEnum: luaCallerArguments[5])),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<TextBox>(boxer: (
      {required TextBox vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedTextBox(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/dart/ui/textBox.ts
import { IRect } from "./rect";
import { TextDirection } from "./textDirection";
declare const dart: {
    ui: {
        textBoxFromLTRBD: (
            left: number,
            top: number,
            right: number,
            bottom: number,
            direction: TextDirection
        ) => ITextBox;
    };
};
export interface ITextBox {
    left: number;
    top: number;
    right: number;
    bottom: number;
    direction: TextDirection;
    toRect: () => IRect;
    getStart: () => number;
    getEnd: () => number;
    getHashCode: () => number;
    toString: () => string;
}
export class TextBox {
    public readonly left: number = undefined as any;
    public readonly top: number = undefined as any;
    public readonly right: number = undefined as any;
    public readonly bottom: number = undefined as any;
    public readonly direction: TextDirection = undefined as any;
    public static fromLTRBD(
        left: number,
        top: number,
        right: number,
        bottom: number,
        direction: TextDirection
    ): ITextBox {
        return dart.ui.textBoxFromLTRBD(left, top, right, bottom, direction);
    }
    private readonly _dart_toRect: () => IRect = undefined as any;
    private readonly _dart_getStart: () => number = undefined as any;
    private readonly _dart_getEnd: () => number = undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    public toRect(): IRect {
        return this._dart_toRect();
    }
    public getStart(): number {
        return this._dart_getStart();
    }
    public getEnd(): number {
        return this._dart_getEnd();
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
    public toString(): string {
        return this._dart_toString();
    }
}
